presto-pinot/presto-pinot-lib/src/main/java/com/facebook/presto/pinot/PinotClusterInfoFetcher.java

```diff
@@ -0,0 +1,394 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.pinot;
+
+import com.facebook.airlift.http.client.HttpClient;
+import com.facebook.airlift.http.client.Request;
+import com.facebook.airlift.http.client.StaticBodyGenerator;
+import com.facebook.airlift.http.client.StringResponseHandler;
+import com.facebook.airlift.json.JsonCodec;
+import com.facebook.airlift.json.JsonCodecBinder;
+import com.facebook.airlift.log.Logger;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Ticker;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.net.HttpHeaders;
+import org.apache.pinot.common.data.Schema;
+
+import javax.inject.Inject;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.facebook.airlift.http.client.StringResponseHandler.createStringResponseHandler;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.apache.pinot.common.config.TableNameBuilder.extractRawTableName;
+
+public class PinotClusterInfoFetcher
+{
+    private static final Logger log = Logger.get(PinotClusterInfoFetcher.class);
+    private static final String APPLICATION_JSON = "application/json";
+    private static final Pattern BROKER_PATTERN = Pattern.compile("Broker_(.*)_(\\d+)");
+
+    private static final String GET_ALL_TABLES_API_TEMPLATE = "tables";
+    private static final String TABLE_INSTANCES_API_TEMPLATE = "tables/%s/instances";
+    private static final String TABLE_SCHEMA_API_TEMPLATE = "tables/%s/schema";
+    private static final String ROUTING_TABLE_API_TEMPLATE = "debug/routingTable/%s";
+    private static final String TIME_BOUNDARY_API_TEMPLATE = "debug/timeBoundary/%s";
+
+    private final PinotConfig pinotConfig;
+    private final PinotMetrics pinotMetrics;
+    private final HttpClient httpClient;
+
+    private final Ticker ticker = Ticker.systemTicker();
+
+    private final LoadingCache<String, List<String>> brokersForTableCache;
+
+    private final JsonCodec<GetTables> tablesJsonCodec;
```

Comment:
You probably don't have to inject it. For simple types, you can directly infer from the annotation:
```
 private final JsonCodec<GetTables> tablesJsonCodec = jsonCodec(GetTables.class);
```